model/coverage.py

```diff
@@ -664,9 +663,15 @@ def bulk_add(
         update = (
             WorkCoverageRecord.__table__.update()
             .where(
-                and_(
-                    WorkCoverageRecord.work_id.in_(work_ids),
-                    WorkCoverageRecord.operation == operation,
+                WorkCoverageRecord.id.in_(
+                    _db.query(WorkCoverageRecord.id)
+                    .with_for_update()
+                    .filter(
+                        and_(
+                            WorkCoverageRecord.work_id.in_(work_ids),
+                            WorkCoverageRecord.operation == operation,
+                        )
+                    )
```

Comment:
Do you need an `order_by(…)` in here, as well?

Reply:
`work_ids` contains already sorted array. The list of `Work` objects was sorted before in `WorkCoverageProvider.items_that_need_coverage` inside `Work.missing_coverage_from`.